public-to-private/run-workflow/main.go

```diff
@@ -0,0 +1,230 @@
+		if timestamp == nil {
+			timestamp = wf.CreatedAt
+		}
+		// we only care about uncompleted workflows
+		if *wf.Status != COMPLETED_STATUS {
+			if currentWorkflow == nil {
+				currentWorkflow = wf
+			}
+			fmt.Printf("found an active workflow\n")
+			// we want the latest workflow so if this one is newer, use it
+			if wf.CreatedAt.After(timestamp.Time) {
+				fmt.Printf("workflow is newer, using it unless we find a newer one\n")
+				currentWorkflow = wf
+			}
+		}
+	}
+	if currentWorkflow == nil {
+		err := fmt.Errorf("failed to find an unfinished workflow")
+		fmt.Printf("%v\n", err)
+		return nil, err
+	}
+	return currentWorkflow, nil
+}
+
+// GetMostRecentWorkflowRunId Check the list of current workflows and get the most recent one that is not complete
+func GetMostRecentWorkflowRunId(client *github.Client, inputs *action.Inputs, workflowStartedAt time.Time) (int64, error) {
+	maxRetrys := 5
+	var err error
+	var workflow *github.WorkflowRun
+	var runs *github.WorkflowRuns
+	for i := 0; i < maxRetrys; i++ {
+		fmt.Printf("Checking for active workflows attempt: %d\n", i+1)
+
+		// if err is nil then it is the first time and we do not need to sleep, otherwise wait
+		if err != nil {
+			time.Sleep(inputs.RetryInterval)
+		}
+
+		// get the list
+		runs, err = GetListOfWorkflowRuns(client, inputs, workflowStartedAt)
+		if err != nil || runs == nil {
+			continue
+		}
+
+		// check the list for the most recent unfinished workflow
+		workflow, err = GetLatestWorkflowRunFromList(runs)
+		if err != nil || workflow == nil {
+			continue
+		}
+
+		// if we make it here then we have found the latest unfinished workflow
+		break
+	}
+
+	// fail if we ended the retry loop in error
+	if err != nil || workflow == nil {
+		err = fmt.Errorf("did not find any active workflows: %v", err)
+		return 0, err
+	}
+
+	fmt.Printf("Successfully found a workflow in progress: %d\n", *workflow.ID)
+	return *workflow.ID, nil
+}
+
+// GetWorkflowRun Gets the workflow run with updated status
+func GetWorkflowRun(client *github.Client, inputs *action.Inputs, workflowID int64) (*github.WorkflowRun, error) {
+	wfr, resp, err := client.Actions.GetWorkflowRunByID(context.Background(), inputs.Owner, inputs.Repository, workflowID)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("Failed to get the workflow, found status code: %d", resp.StatusCode)
+	}
+	return wfr, nil
+}
+
+// PollWorkflow Poll the workflow until the status is complete or we hit the timeout
+func PollWorkflow(client *github.Client, inputs *action.Inputs, workflowID int64) *action.Outputs {
+	fmt.Printf("Beginning to poll for the workflows status every %v until it is complete or we timeout after %v\n", inputs.RetryInterval, inputs.Timeout)
+	pollStart := time.Now()
+	stop := false
+	var status string
+	var latestWorkflow *github.WorkflowRun
+	var err error
+	testContext, testCancel := context.WithTimeout(context.Background(), inputs.Timeout)
```

Comment:
Between two polls you can actually get different workflow runs for the same workflowID, if the triggered workflow has a re-run executed, it'll keep the same workflowRunID but increment its attempt number instead. If this is the behaviour you want (always report status of the most recent re-run of the triggered workflow, then that's fine). If you want the behaviour to be pinned to the original workflow run detected, you should instead use https://docs.github.com/en/rest/actions/workflow-runs#get-a-workflow-run-attempt

Reply:
Switched to using attempt and forced it to 1 for now. If we need to allow for later attempts we can add that later.